pkg/urbit/noun/manage.c

```diff
@@ -1713,70 +1718,136 @@ _cm_signals(void)
 # endif
 }
 
+/* _cm_malloc_ssl(): openssl-shaped malloc
+*/
+static void*
+_cm_malloc_ssl(size_t len_i
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+               , const char* file, int line
+#endif
+               )
+{
+  return u3a_malloc(len_i);
+}
+
+/* _cm_realloc_ssl(): openssl-shaped realloc.
+*/
+static void*
+_cm_realloc_ssl(void* lag_v, size_t len_i
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+                , const char* file, int line
+#endif
+                )
+{
+  return u3a_realloc(lag_v, len_i);
+}
+
+/* _cm_free_ssl(): openssl-shaped free.
+*/
+static void
+_cm_free_ssl(void* tox_v
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+             , const char* file, int line
+#endif
+             )
+{
+  return u3a_free(tox_v);
+}
+
 extern void u3je_secp_init(void);
-extern void u3je_secp_stop(void);
 
+/* _cm_crypto(): initialize openssl and crypto jets.
+*/
 static void
 _cm_crypto()
 {
   /* Initialize OpenSSL with loom allocation functions. */
-  if ( 0 == CRYPTO_set_mem_functions(&u3a_malloc_ssl,
-                                     &u3a_realloc_ssl,
-                                     &u3a_free_ssl) ) {
+  if ( 0 == CRYPTO_set_mem_functions(&_cm_malloc_ssl,
+                                     &_cm_realloc_ssl,
+                                     &_cm_free_ssl) ) {
     u3l_log("%s\r\n", "openssl initialization failed");
     abort();
   }
 
   u3je_secp_init();
 }
 
+/* _cm_realloc2(): gmp-shaped realloc.
+*/
+static void*
+_cm_realloc2(void* lag_v, size_t old_i, size_t new_i)
+{
+  return u3a_realloc(lag_v, new_i);
+}
+
+/* _cm_free2(): gmp-shaped free.
+*/
+static void
+_cm_free2(void* tox_v, size_t siz_i)
+{
+  return u3a_free(tox_v);
+}
+
 /* u3m_init(): start the environment.
 */
 void
-u3m_init(void)
+u3m_init(size_t len_i)
 {
   _cm_limits();
   _cm_signals();
   _cm_crypto();
 
-  /* Make sure GMP uses our malloc.
-  */
-  mp_set_memory_functions(u3a_malloc, u3a_realloc2, u3a_free2);
+  //  make sure GMP uses our malloc.
+  //
+  mp_set_memory_functions(u3a_malloc, _cm_realloc2, _cm_free2);
 
-  /* Map at fixed address.
-  */
+  //  make sure that [len_i] is a fully-addressible non-zero power of two.
+  //
+  if (  !len_i
```

Comment:
Nit, but can we change `!len_i` to `0 != len_i`.

Reply:
We could (well, `0 == len_i`), but this is idiomatic C. (I'm responsible for many of the explicit comparisons to zero in this codebase -- those stemmed from a misplaced fear of loobeans. I regret the error.)

Reply:
Fair, but it's slightly more obvious to use `0 == len_i` since we're compare `len_i` to `1 << (u3a_page + 2)` and `u3a_bytes` anyway, so the pattern of "compare `len_i` to some number using a binary relation" is already established. Like I said though, it's a nit, and it doesn't matter much either way.

Reply:
The second conditional is actually `(len_i & (len_i - 1))` -- "if any bits other than the most significant are set". The whole conditional is checking for violations of our map-size requirements -- size is zero, or not a power of two, or less than our page size, or greater than our compile-time maximum.

Reply:
There are many things that are hard in C, and some of them are avoidable. I'm don't think it's reasonable to place the standard boolean conversions are in either set (which is why I said I regret the error).

Loobeans are another story, but I think the solution is just to quarantine them to the interpreter and relevant jets.